GitUp/Application/CloneWindowController.m

```diff
@@ -0,0 +1,109 @@
+//
+//  CloneWindowController.m
+//  Application
+//
+//  Created by Dmitry Lobanov on 08.10.2019.
+//
+
+#import "CloneWindowController.h"
+#import <GitUpKit/GitUpKit.h>
+
+@interface CloneWindowControllerResult ()
+- (instancetype)initWithRepositoryURL:(NSURL *)url directoryPath:(NSString *)path recursive:(BOOL)recursive;
+@end
+
+@implementation CloneWindowControllerResult
+- (instancetype)initWithRepositoryURL:(NSURL *)url directoryPath:(NSString *)path recursive:(BOOL)recursive {  
+  if ((self = [super init])) {
+    self.repositoryURL = url;
+    self.directoryPath = path;
+    self.recursive = recursive;
+  }
+  return self;
+}
+
+- (BOOL)invalidRepository {
+  return self.repositoryURL == nil;
+}
+
+- (BOOL)emptyDirectoryPath {
+  return self.directoryPath == nil;
+}
+@end
+
+@interface CloneWindowController ()
+@property(nonatomic, weak) IBOutlet NSTextField* urlTextField;
+@property(nonatomic, weak) IBOutlet NSButton* cloneRecursiveButton;
+@property (nonatomic, assign, readonly) BOOL urlExists;
+@property (nonatomic, assign, readonly) BOOL recursive;
+@end
+
+@implementation CloneWindowController
+#pragma mark - Initialization
+- (instancetype)init {
+  return [super initWithWindowNibName:@"CloneWindowController"];
+}
+
+#pragma mark - Window Lifecycle
+- (void)beforeRunInModal {
+  self.urlTextField.stringValue = self.url;
+  self.cloneRecursiveButton.state = NSOnState;
+}
+
+- (void)windowDidLoad {
+  [super windowDidLoad];
+  [self beforeRunInModal];
+}
+
+#pragma mark - Actions
+- (IBAction)dismissModal:(id)sender {
+  [NSApp stopModalWithCode:[(NSButton *)sender tag]];
+  [self close];
+}
+
+#pragma mark - Modal
+- (void)runModalForURL:(NSString *)url completion:(nonnull void (^)(CloneWindowControllerResult * _Nonnull))completion {
+  if (!completion) {
+    return;
+  }
+  
+  self.url = url;
+  if (self.windowLoaded) {
+    [self beforeRunInModal];
```

Comment:
Wouldn't this already have been run in `windowDidLoad`?

Reply:
It is part of "restoration cycle".

1. Window is loaded, so, all outlets are set and we can (re)set them correctly.
2. Window is not loaded, so, we skip if-condition and wait for next line where window will be loaded.

```objective-c
  if (self.windowLoaded) { // We have outlets, so, we can set UI elements
    [self beforeRunInModal];
  }
  // If not, we will skip windowLoaded if-condition and wait for `self.window` statement below. Check -windowDidLoad method for details.
  if ([NSApp runModalForWindow:self.window] && self.urlExists) {
```

We could remove `self.windowLoaded` check  and use objective-c nil-chaining feature.
```objective-c
//TIP: We don't care if window is loaded or not. This method modifies outlets. If window is not loaded, we invoke setters on nil outlets.
[self beforeRunInModal];
if ([NSApp runModalForWindow:self.window] && self.urlExists) {
```

Reply:
@lucasderraugh What do you think about it?